daemon/src/engine/framework/Crypto.cpp

```diff
@@ -0,0 +1,289 @@
+{
+    base64_decode_ctx ctx;
+    nettle_base64_decode_init( &ctx );
+
+    Data temp( BASE64_DECODE_LENGTH( input.size() ) );
+    std::size_t decoded_bytes = 0;
+    if ( !nettle_base64_decode_update( &ctx, &decoded_bytes, temp.data(),
+                                       input.size(), input.data() ) )
+    {
+        return false;
+    }
+
+    if ( !nettle_base64_decode_final( &ctx ) )
+    {
+        return false;
+    }
+
+    temp.erase( temp.begin() + decoded_bytes, temp.end() );
+    output = temp;
+    return true;
+}
+
+
+} // namespace Encoding
+
+
+namespace Hash {
+
+Data Sha256( const Data& input )
+{
+    sha256_ctx ctx;
+    nettle_sha256_init( &ctx );
+    nettle_sha256_update( &ctx, input.size(), input.data() );
+    Data output( SHA256_DIGEST_SIZE  );
+    nettle_sha256_digest( &ctx, SHA256_DIGEST_SIZE, output.data() );
+    return output;
+}
+
+Data Md5( const Data& input )
+{
+    md5_ctx ctx;
+    nettle_md5_init( &ctx );
+    nettle_md5_update( &ctx, input.size(), input.data() );
+    Data output( MD5_DIGEST_SIZE );
+    nettle_md5_digest( &ctx, MD5_DIGEST_SIZE, output.data() );
+    return output;
+}
+
+
+} // namespace Hash
+
+/*
+ * Adds PKCS#7 padding to the data
+ */
+void AddPadding( Data& target, std::size_t block_size )
+{
+    auto pad = block_size - target.size() % block_size;
+    target.resize( target.size() + pad, pad );
+}
+
+/*
+ * Encrypts using the AES256 algorthim
+ * PRE: Key is 256-bit (32 octects) long
+ * POST: output contains the encrypted data
+ * Notes: plain_text will be padded using the PKCS#7 algorthm,
+ *        if the decoding fails, output is unchanged
+ * Returns true on success
+ */
+bool Aes256Encrypt( Data plain_text, const Data& key, Data& output )
+{
+    if ( key.size() != AES256_KEY_SIZE )
+    {
+        return false;
+    }
+    aes256_ctx ctx;
+    nettle_aes256_set_encrypt_key( &ctx, key.data() );
+    AddPadding( plain_text, AES_BLOCK_SIZE );
+    output.resize( plain_text.size(), 0 );
+    nettle_aes256_encrypt( &ctx, plain_text.size(),
+        output.data(), plain_text.data() );
+    nettle_aes256_decrypt( &ctx, plain_text.size(),
+        plain_text.data(), output.data() );
+    return true;
+}
+
+/*
+ * Encrypts using the AES256 algorthim
+ * PRE: Key is 256-bit (32 octects) long,
+ *      cypher_text.size() is an integer multiple of the AES block size
+ * POST: output contains the decrypted data
+ * Note: If the decoding fails, output is unchanged
+ * Returns true on success
+ */
+bool Aes256Decrypt( Data cypher_text, const Data& key, Data& output )
+{
+    if ( key.size() != AES256_KEY_SIZE || cypher_text.size() % AES_BLOCK_SIZE )
+    {
+        return true;
+    }
+
+    aes256_ctx ctx;
+    nettle_aes256_set_decrypt_key( &ctx, key.data() );
+    output.resize( cypher_text.size(), 0 );
+    nettle_aes256_decrypt( &ctx, cypher_text.size(),
+        output.data(), cypher_text.data() );
+
+    // Strip the PKCS#7 padding
+    if ( !output.empty() )
```

Comment:
Like it's done below.

According to PKCS#7 you always add at least one byte, and all of the padding bytes have as value the number of padding bytes.